File: Algorithm/Leetcode/[E]1897. Redistribute Characters to Make All Strings Equal.py
from collections import Counter
class Solution:
    def makeEqual(self, words: List[str]) -> bool:
        d = []
        
        for i in range(len(words)):
            d.append(Counter(words[i]))

        da  = Counter()
        for i in range(len(d)):
            da += d[i]
                
        d_list = list(da.items())
        #print(d_list)
        
        n = len(words)
        
        for k, v in d_list:
            #print (k,v)
            if (v%n) != 0:
                return False
        return True
    
# Counter를 사용해서 빼내는법, for문을 통해 딕셔너리에 넣고 빼내는법
# 둘중하나를 사용해서 갯수를 구하면 된다.
# 다음과정에서 생각을 잘못해서 해답을 봤는데
# 각 글자당 갯수가 words의 갯수로 나누어 떨어져야 정답이 나올수있다. 
# 특이하게 배운것
# Counter로 뽑아낸 것은 Counter 객체에만 제대로 담긴다 list dict 이상하게 넣어보려하면 다 실패한다.
# Counter로 뽑아낸 것은 딕셔너리형태지만 합교집합 가능, 그리고 덧셈 뺄셈도 가능
# most_common이나 elements 같은 메소드 들도 가지고 있음

File: Algorithm/Leetcode/[E]1897. Redistribute Characters to Make All Strings Equal.java
class Solution {
    public boolean makeEqual(String[] words) {
        Map<Character, Integer> m  = new HashMap<>();
        
        for(int i = 0; i <words.length; i++){
            for(int j = 0; j < words[i].length(); j++){
                char c =  words[i].charAt(j);
                m.put(c, m.getOrDefault(c, 0) + 1);
            }
        }
               
        
        for (char key: m.keySet()){
            if(m.get(key) % words.length != 0)
                return false;
        }
        return true;
    }
}

/* 
파이썬으로 푼 방식과 동일하게 풀었다. 
9ms
다만 자바로 푸는 방식은 빠르게 푼 사람들이 존재해서 그사람들 답을 보니
배열에다가 증가 때려넣고 마지막에 비교하는 형식으로 풀었음;;
해시맵을 사용하지 않았음
1차적으로 모든 단어의 길이를 더한 값과 단어 갯수가 나누어 떨어지는지 보고
위 방식을 배열에 적용하여 배열값을 단어 갯수와 나눠봄

charAt(index)
put(K,V)
getOrDefault(K, {default_value}):
keySet() iter할때 사용
get(K) key에 따른 Value가져옴 

toCharArray(): String을 char배열로 바꿔준다.


*/